<a href="https://colab.research.google.com/github/seongchanc/react/blob/main/%08ML_%EC%84%B1%EC%B0%AC%ED%98%81_%ED%94%84%EB%A1%9C%EC%A0%9D%ED%8A%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1.실험 내용에 대한 전체 요약**

여러 classification 모델을 설계하여 어떤 모델이 가장 정확한 결과를 도출하는지 비교해보고 어떠한 feature를 통해 모델을 훈련시키고 테스트하는 것이 가장 좋은 모델로 도출되는지에 대해서 결과를 비교해 보고자합니다. 이번 실험에 사용할 알고리즘은 Decision Tree, Multi-Layer Perceptron, LogisticRegression , K-Nearest Neighbor, Random Forest 총 5가지이며 이 기법들을 사용하여 최고의 성능을 나타내는 모델을 찾는 실험을 진행하도록 하겠습니다. 또한 Feature selection 기법을 적용하여 feature를 선택하여 진행할 예정입니다.

# **2. 데이터 출처 및 선정한 데이터에 대한 설명**

### 2.1. Data Description 

데이터 URL: http://archive.ics.uci.edu/ml/datasets/Adult

Adult 데이터셋은 14개의 feature와 약 5만개의 instance를 가지고 있는 데이터셋으로, 데이터셋의 제한 조건인 최소 10개 이상의 feature, 최소 5000개 이상의 instances를 가진 데이터셋이며 fatures의 이해가 쉬운편에 속하며, 많은 양의 데이터를 통해서 결과를 얻어낼 수 있다는 장점때문에 Adult 데이터셋을 선택하게 되었습니다. 

### 2.2. Feature Description 

**Input**

* age: 나이
* workclass: 고용 형태
* fnlwgt: 일련의 관측 결과를 바탕으로 인구조사국이 부여하는 개인의 가중치
* education: 최종 학력
* education-num: 교육 수준을 숫자로 범주화하여 열거한 것.
* marital-status: 결혼 여부
* occupation: 직업
* relationship: 가정 내 각 개인의 관계
* race: 인종
* sex: 성별
* capital-gain: 개인의 자본 이익 (달러)
* capital-loss: 개인의 자본 손실 (달러)
* hours-per-week: 주당 근무시간
* native-country: 출신 국가

**Output**

* income: 연간 소득이 50,000$를 넘기는지, 넘기지 못하는지 여부

# **3. 실험 설계 및 방법**

### 3.1. 선정 모델 및 모델 선정에 대한 이유

선정 모델: Decision Tree, Multi-Layer Perceptron, LogisticRegression , K-Nearest Neighbor, Random Forest 
선정 이유: 해당 모델들을 선택한 이유는 실습을 진행했을때 가장 이해가 잘됐으며 실제로 사용해보고싶은 모델을 선정하게 되었습니다. 



### 3.2. 데이터 셋 전처리

In [ ]:
#data 불러오기 및 확인
import pandas as pd

data = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data', names=['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', 'income'])
data

age          workclass  fnlwgt    education  education-num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32556   27            Private  257302   Assoc-acdm             12   
32557   40            Private  154374      HS-grad              9   
32558   58            Private  151910      HS-grad              9   
32559   22            Private  201490      HS-grad              9   
32560   52       Self-emp-inc  287927      HS-grad              9   

            marital-status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32556   Married-civ-spouse        Tech-support            Wife   White   
32557   Married-civ-spouse   Machine-op-inspct         Husband   White   
32558              Widowed        Adm-clerical       Unmarried   White   
32559        Never-married        Adm-clerical       Own-child   White   
32560   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital-gain  capital-loss  hours-per-week  native-country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32556   Female             0             0              38   United-States   
32557     Male             0             0              40   United-States   
32558   Female             0             0              40   United-States   
32559     Male             0             0              20   United-States   
32560   Female         15024             0              40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32556   <=50K  
32557    >50K  
32558   <=50K  
32559   <=50K  
32560    >50K  

[32561 rows x 15 columns]

In [ ]:
#Data 전처리
data = data[['age', 'fnlwgt', 'education-num', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'race', 'income']]
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

data.sex = le.fit_transform(data.sex)
data.income = le.fit_transform(data.income)

def onehot(data, feature):
  return pd.concat([data, pd.get_dummies(data[feature], prefix=feature)], axis=1).drop([feature], axis=1)

data = onehot(data, 'race')

data

/usr/local/lib/python3.8/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


age  fnlwgt  education-num  sex  capital-gain  capital-loss  \
0       39   77516             13    1          2174             0   
1       50   83311             13    1             0             0   
2       38  215646              9    1             0             0   
3       53  234721              7    1             0             0   
4       28  338409             13    0             0             0   
...    ...     ...            ...  ...           ...           ...   
32556   27  257302             12    0             0             0   
32557   40  154374              9    1             0             0   
32558   58  151910              9    0             0             0   
32559   22  201490              9    1             0             0   
32560   52  287927              9    0         15024             0   

       hours-per-week  income  race_ Amer-Indian-Eskimo  \
0                  40       0                         0   
1                  13       0                         0   
2                  40       0                         0   
3                  40       0                         0   
4                  40       0                         0   
...               ...     ...                       ...   
32556              38       0                         0   
32557              40       1                         0   
32558              40       0                         0   
32559              20       0                         0   
32560              40       1                         0   

       race_ Asian-Pac-Islander  race_ Black  race_ Other  race_ White  
0                             0            0            0            1  
1                             0            0            0            1  
2                             0            0            0            1  
3                             0            1            0            0  
4                             0            1            0            0  
...                         ...          ...          ...          ...  
32556                         0            0            0            1  
32557                         0            0            0            1  
32558                         0            0            0            1  
32559                         0            0            0            1  
32560                         0            0            0            1  

[32561 rows x 13 columns]

In [ ]:
# train data, test data 구별
x = data.drop(columns=['income'])
y = data[['income']]

from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=42)

# SelectKBest와 chi2 테스트를 사용하여 adult에서 두개의 가장 중요한 특성을 선택합니다. Feature selection 기법 함수
from sklearn.feature_selection import SelectKBest, chi2

def k_new(x, y):
  x_new = SelectKBest(chi2, k=3).fit_transform(x, y)
  
  return x_new

print(k_new(x,y))


[[ 77516   2174      0]
 [ 83311      0      0]
 [215646      0      0]
 ...
 [151910      0      0]
 [201490      0      0]
 [287927  15024      0]]


### 3.3. 기본 모델 생성, 훈련 및 성능 평가

### **Decision Tree**

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

dt_model = DecisionTreeClassifier(random_state=42)

cross_val_score(dt_model, x_train, y_train).mean()

0.7781675022649267

In [ ]:
dt_model_1 = DecisionTreeClassifier(max_depth = 10, random_state=42)

cross_val_score(dt_model_1, x_train, y_train).mean()

0.8361707902464698

In [ ]:
dt_model_2 = DecisionTreeClassifier(max_depth = 10, max_leaf_nodes= 30, random_state=42)

cross_val_score(dt_model_2, x_train, y_train).mean()

0.8390228125014136

In [ ]:
dt_model_3 = DecisionTreeClassifier(min_samples_split = 20, random_state=42)

cross_val_score(dt_model_3, x_train, y_train).mean()

0.8041860118627971

In [ ]:
dt_model_4 = DecisionTreeClassifier(max_depth = 10, max_leaf_nodes= 20,min_samples_split = 10,random_state=42)

cross_val_score(dt_model_4, x_train, y_train).mean()

0.8373554818925693

**adult 데이터셋을 이용하여 Decision tree 알고리즘을 사용해봤습니다. 모델에 대해서 훈련을 해본 결과 가장 기본적으로 제공되는 값을 사용한 모델인 dt_model 모델은 평균적으로 낮은 성능을 보여주고 있습니다. 가장 높은 성능을 보여준 모델은 max_depth = 10, max_leaf_nodes= 30로 파라미터를 조종해준 dt_model_2 모델이 가장 좋은 성능을 보여준 것으로 나타났습니다. max_depth와 max_leaf_nodes를 설정했을때 가장 좋은 성능을 보여줬기 때문에 min_samples_split를 추가적으로 수정한다면 더 좋은 성능을 보여줄것이라 생각하고 실험을 진행해봤지만 결과는 더 낮은 성능을 보여주었습니다. 그러므로 최적의 모델은 dt_model_2 모델입니다.**

### **Decision Tree에 Feature selection 기법 적용**

In [ ]:
x_new = k_new(x_train, y_train)

cross_val_score(dt_model, x_new, y_train).mean()

0.7594335917141555

In [ ]:
x_new_1 = k_new(x_train, y_train)

cross_val_score(dt_model_1, x_new_1, y_train).mean()

0.8280100281415879

In [ ]:
x_new_2 = k_new(x_train, y_train)

cross_val_score(dt_model_2, x_new_2, y_train).mean()

0.8268692635131162

In [ ]:
x_new_3 = k_new(x_train, y_train)

cross_val_score(dt_model_3, x_new_3, y_train).mean()

0.7823800108585586

In [ ]:
x_new_4 = k_new(x_train, y_train)

cross_val_score(dt_model_4, x_new_4, y_train).mean()

0.8262989052605685

**그 이후 Feature selection기법을 사용하여 동일하게 테스트를 진행해봤습니다. 
Feature selection 기법을 사용했을 때 사용하지 않았을 때보다 성능이 낮아지는 것을 확인할 수 있었습니다. 이러한 결과가 나온 이유는 기존의 Decision tree 알고리즘으로 충분히 최적의 파라미터를 사용하여 진행하기 때문이라고 생각됩니다. 수업에서도 교수님이 언급하신 내용으로 Decision tree는 Feature selection 기법과 큰 차이가 없다고 말씀하셨던 것이 이번 실험 결과로 증명할 수 있을것입니다.** 

### **Decision Tree: test set에 대한 최종 평가**

In [ ]:
dt_model.fit(x_train,y_train)
dt_model.score(x_test,y_test)

0.7749001944927832

In [ ]:
dt_model_1.fit(x_train,y_train)
dt_model_1.score(x_test,y_test)

0.8388780837342614

In [ ]:
dt_model_2.fit(x_train,y_train)
dt_model_2.score(x_test,y_test)

0.8385709898659024

In [ ]:
dt_model_3.fit(x_train,y_train)
dt_model_3.score(x_test,y_test)

0.808475790766711

In [ ]:
dt_model_4.fit(x_train,y_train)
dt_model_4.score(x_test,y_test)

0.8366260620329614

**위 실험에서 사용했던 모델들을 통해 최종적으로 test set에 대하여 score를 나타냈을 때 dt_model_1 모델이 가장 높은 성능을 보였습니다. dt_model_1 모델에서 파라미터 값은 max_depth = 10으로 설정하였고, 성능이 조금 낮게 나온 dt_model_2 모델은 max_leaf_nodes 파라미터를 30으로 설정하였습니다. max_leaf_nodes는 말단 노드의 최대 개수를 나타내며 데이터 셋을 생각해 봤을때 30개로 설정하는 것과 default 값으로 설정되는 것에서 차이가 나지 않을것이라고 예측했습니다.**

### **Multi-Layer Perceptron**

In [ ]:
from sklearn.neural_network import MLPClassifier

mlp_model = MLPClassifier(random_state=42)
cross_val_score(mlp_model, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when 

0.6817648016195633

In [ ]:
mlp_model_1 = MLPClassifier(learning_rate_init = 0.01, max_iter = 150)
cross_val_score(mlp_model_1, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when 

0.8025622906572988

In [ ]:
mlp_model_2 = MLPClassifier(learning_rate_init = 0.002, max_iter= 1000)
cross_val_score(mlp_model_2, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when 

0.6793516934279156

In [ ]:
mlp_model_3 = MLPClassifier(random_state=42, max_iter=1000, learning_rate_init=0.01)
cross_val_score(mlp_model_3, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when 

0.8032203681996496

In [ ]:
mlp_model_4 = MLPClassifier(hidden_layer_sizes=(70,),max_iter=1000, learning_rate_init=0.01)
cross_val_score(mlp_model_4, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when 

0.8050192873678738

**MLP를 사용했을때 모델들은 Decision Tree를 사용한 모델보다 낮은 성능을 보여주는 것으로 확인되었습니다. 평균적으로 성능 수치가 매우 낮습니다. 가장 높은 성능을 보여준 모델은 mlp_model_4 모델입니다. 0.8050192873678738의 score로 파라미터중 하나인 learning_rate_init 값을 적당한 값으로 설정했을때 높은 성능을 보여주는 것을 확인할 수 있습니다. mlp_model_4와 비슷한 성능을 보인 mlp_model_1또 한 learning_rate_init 파라미터를 동일한 값인 0.01로 설정했습니다. max_iter 차이가 심하게 수정했지만 성능에서 별차이가 나타나지 않을것으로 보아 max_iter의 중요도는 낮다고 보여집니다.**

### **Multi-Layer Perceptron에 Feature selection 기법 적용**

In [ ]:
x_new = k_new(x_train, y_train)

cross_val_score(mlp_model, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when 

0.6753151335216754

In [ ]:
cross_val_score(mlp_model_1, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when 

0.7989207274214023

In [ ]:
cross_val_score(mlp_model_2, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when 

0.6788492083849015

In [ ]:
cross_val_score(mlp_model_3, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when 

0.801202275927696

In [ ]:
cross_val_score(mlp_model_4, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when 

0.7978238609365329

**Feature selection을 적용했을때, 전체적인 성능이 낮아지는 것을 확인할 수 있습니다. MLP에서 Feature를 선택하는것이 모델에 영향을 많이 주기때문에 적절한 Feature를 선택하지 못한다면 성능이 낮아진다는 것을 확인할 수 있었습니다. 그렇기 때문에 다음에 기회가 된다면 다른 Feature selection을 사용하여 성능이 증가되는지 확인해보고싶습니다.**

### **MLP: test set에 대한 최종 평가**

In [ ]:
mlp_model.fit(x_train, y_train)
mlp_model.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.7946565666905517

In [ ]:
mlp_model_1.fit(x_train, y_train)
mlp_model_1.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.8083734261439247

In [ ]:
mlp_model_2.fit(x_train, y_train)
mlp_model_2.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.2368717371276487

In [ ]:
mlp_model_3.fit(x_train, y_train)
mlp_model_3.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.8110349063363701

In [ ]:
mlp_model_4.fit(x_train, y_train)
mlp_model_4.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:1109: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.7993653393387246

**위 실험에서 사용했던 모델들을 통해 최종적으로 test set에 대하여 score를 나타냈을 때 mlp_model_3가 가장 높은 성능을 보였습니다. mlp_model_3 모델은 cross_val_socre에서 0.8032203681996496의 점수로 mlp_model_4에 이어 2번째로 높은 성능을 보였습니다. 하지만 test_set에 적용해본 결과 mlp_model_4보다 좋은 성능인 0.8110349063363701의 score 값을 도출해낸것을 확인할 수 있었습니다. 또한 learning_rate_init를 과도하게 적게 설정한 mlp_model_2 모델은 압도적으로 낮은 성능을 보여주었습니다. 이것을 통해 learning_rate_init을 낮게 설정하게 되면 overfitting되어 test set에서 안좋은 결과가 나타난다는 점을 알 수 있습니다.**


### **LogisticRegression**

In [ ]:
from sklearn.linear_model import LogisticRegression

lr_model = LogisticRegression()
cross_val_score(lr_model, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example usin

0.7951912139034978

In [ ]:
lr_model_1 = LogisticRegression(solver='sag' ,max_iter=1000)
cross_val_score(lr_model_1, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a

0.7917691125115869

In [ ]:
lr_model_2 = LogisticRegression(solver='sag' ,max_iter=500)
cross_val_score(lr_model_2, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a

0.7905844593641882

In [ ]:
lr_model_3 = LogisticRegression(solver='saga' ,max_iter=500)
cross_val_score(lr_model_3, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a

0.7908038345860972

In [ ]:
lr_model_4 = LogisticRegression(solver='saga' ,max_iter=1000)
cross_val_score(lr_model_4, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a

0.7905844593641882

**LogisiticRegression을 사용한 모델들은 파라미터를 변경해도 성능에서 큰 차이를 보이지 않았습니다. 가장 좋은 성능을 보인것은 파라미터를 수정하지않고 default 값으로 설정한 기본 모델이 가장 좋은 성능을 보였습니다. max_iter의 default 값은 100이며 solver의 default 값은 lbfgs 입니다. LogisiticRegression 모델은 어떤 파라미터를 수정하더라도 성능에 큰 영향을 미치지 않았기 때문에 가장 까다로운 모델들 중 하나였습니다. 실험을 진행하면서 이러한 결과가 나온 이유를 정확하게 파악하지 못했지만 max_iter 값이 기본값인 100으로 충분하기 때문에 그 이상으로 설정하는 것은 너무 크게 설정한게 문제라고 생각됩니다.**

### **LogisiticRegression에 Feature selection 기법 적용**

In [ ]:
x_new = k_new(x_train, y_train)

cross_val_score(lr_model, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example usin

0.7980436211454498

In [ ]:
cross_val_score(lr_model_1, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a

0.7917691125115869

In [ ]:
cross_val_score(lr_model_2, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a

0.7905405804699364

In [ ]:
cross_val_score(lr_model_3, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a

0.7908038345860972

In [ ]:
cross_val_score(lr_model_4, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a

0.7905405804699364

**LogisiticRegression 모델에 Feature selection을 적용했을때의 성능차이는 크게 나타나지 않았습니다. 가장 높은 성능을 보인 모델은 Feature selection을 적용하기전에서 가장 좋은 성능을 보였던 lr_model입니다. 0.7980436211454498의 score로 다른 모델들 중 가장 높은 성능을 보였습니다. 하지만 다른 모델들과도 큰 차이가 없기 때문에 가장 좋은 모델이라고 하기에는 애매한 부분이 있습니다. 여전히 왜 lr_model이 가장 좋은 성능을 보이는지 정확하게 파악하지 못했기 때문에 연구를 더욱 진행해보는것이 필요해 보입니다. 차이가 나지않는 점에대해 예측해보자면 데이터 셋의 각 feature의 분포가 잘되어있으므로 확률로 예측하는 회귀 알고리즘에서 비슷한 결과가 나오는 것으로 예측해볼 수 있습니다.**

### **LogisiticRegression: test set에 대한 최종 평가**

In [ ]:
lr_model.fit(x_train, y_train)
lr_model.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.8022315487767427

In [ ]:
lr_model_1.fit(x_train, y_train)
lr_model_1.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


0.7992629747159382

In [ ]:
lr_model_2.fit(x_train, y_train)
lr_model_2.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


0.7984440577336472

In [ ]:
lr_model_3.fit(x_train, y_train)
lr_model_3.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


0.7980345992425018

In [ ]:
lr_model_4.fit(x_train, y_train)
lr_model_4.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/utils/validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_sag.py:352: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


0.7984440577336472

**위 실험에서 사용했던 모델들을 통해 최종적으로 test set에 대하여 score를 나타냈을 때 lr_model 모델이 가장 높은 성능을 보였습니다. lr_model 모델은 cross_val_socre에서 0.7951912139034978의 점수로 lr 모델들 중 가장 좋은 성능을 보였습니다. test set에 적용했을때도 마찬가지로 lr_model이 가장 좋은 성능을 나타낸것을 확인할 수 있었습니다.**

### **K-Nearest Neighbor**

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
knn_model = KNeighborsClassifier()

cross_val_score(knn_model, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

0.7700948444368559

In [ ]:
knn_model_1 = KNeighborsClassifier(n_neighbors = 3)

cross_val_score(knn_model_1, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

0.7540807227284106

In [ ]:
knn_model_2 = KNeighborsClassifier(n_neighbors = 3, weights = "distance")

cross_val_score(knn_model_2, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

0.7468412152846386

In [ ]:
knn_model_3 = KNeighborsClassifier(n_neighbors = 20)

cross_val_score(knn_model_3, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

0.7960249802670096

In [ ]:
knn_model_4 = KNeighborsClassifier(n_neighbors = 20, weights = "distance")

cross_val_score(knn_model_4, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

0.7914181583549735

**KNN 모델에서는 n_neighbors 설정이 모델 정확도를 결정하는데 가장 큰 역할을 하는것을 알 수 있습니다.n_neighbors은 k의 개수를 의미합니다. knn_model의 n_neighbors 값은 default 값인 5로 설정되었으며, knn_model_1에서 n_neighbors의 값을 3으로 감소시켰을때 성능이 저하되는 것을 확인할 수 있습니다. 또한 knn_model_3에서 n_neighbors의 값을 20으로 설정했을때 가장 좋은 성능을 보여준것을 통해 n_neighbors 값을 높게 설정하는것이 중요하다는것을 알 수 있습니다. 하지만 n_neighbors의 값을 100으로 설정하게되면 오히려 성능이 감소하는 것을 확인했기 때문에 적당한 양의 n_neighbors 값을 설정해주는 것이 성능에 좋습니다. 또한 x 데이터를 분류를 할 때 k개의 이웃 중 거리가 가까운 이웃의 영향을 더 많이 받도록 가중치를 설정하려면 weights = "distance"로 설정하면 되지만 이것을 distance로 설정했을시 default값인 'uniform'보다 낮은 성능을 보였습니다.**

### **K-Nearest Neighbor에 Feature selection 기법 적용**

In [ ]:
x_new = k_new(x_train, y_train)

cross_val_score(knn_model, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

0.7705335467572977

In [ ]:
cross_val_score(knn_model_1, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

0.7509654174832802

In [ ]:
cross_val_score(knn_model_2, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

0.7362234853432078

In [ ]:
cross_val_score(knn_model_3, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

0.7963321036527472


In [ ]:
cross_val_score(knn_model_4, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)
/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for exam

0.7653563954667395

**Feature selection을 적용했을때 knn_model_3 모델의 성능이 증가한것을 확인할 수 있습니다. Feature selection을 적용했을 때 성능이 증가한 모델의 특징은 k의 개수를 나타내는 n_neighbors 값을 충분히 증가시킨 모델들이 증가했다는 점입니다. 너무 작은 n_neighbors를 설정한 모델들은 Feature selection을 적용했을때 성능이 감소했으며 n_neighbors를 적당한 값으로 설정한 모델들은 성능이 증가했습니다. Feature selection을 통해 선별한 Feature에 k개의 갯수가 20이 적당하다는 결과를 알 수 있습니다.**


### **K-Nearest Neighbor: test set에 대한 최종 평가**

In [ ]:
knn_model.fit(x_train, y_train)
knn_model.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


0.7783805916675197

In [ ]:
knn_model_1.fit(x_train, y_train)
knn_model_1.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


0.7565769270140239

In [ ]:
knn_model_2.fit(x_train, y_train)
knn_model_2.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


0.7514586958747057

In [ ]:
knn_model_3.fit(x_train, y_train)
knn_model_3.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


0.8042788412324701

In [ ]:
knn_model_4.fit(x_train, y_train)
knn_model_4.score(x_test, y_test)

/usr/local/lib/python3.8/dist-packages/sklearn/neighbors/_classification.py:198: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return self._fit(X, y)


0.7949636605589109

**test set에 적용해본 결과 train set에 적용한 결과와 같은 값을 나타냈습니다. knn_model_3 모델이 0.8042788412324701이란 score로 가장 좋은 성능을 보였습니다.knn 모델의 핵심은 k 개수를 나타내는 n_neighbors 값을 적절히 설정하는 것이며 weights 파라미터의 값은 default 값인 'uniform'으로 설정하는 것이 모델 성능에 좋은 영향을 미치는 것을 확인할 수 있었습니다.**

### **Random Forest**




In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(random_state = 42)

cross_val_score(rf_model, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 

0.8220428161376159

In [ ]:
rf_model_1 = RandomForestClassifier(n_estimators= 200, random_state = 42)

cross_val_score(rf_model_1, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 

0.8229202496525252

In [ ]:
rf_model_2 = RandomForestClassifier(n_estimators=5, random_state = 42)

cross_val_score(rf_model_2, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 

0.8098018077257461

In [ ]:
rf_model_3 = RandomForestClassifier(n_estimators= 200, min_samples_split = 10 ,random_state = 42)

cross_val_score(rf_model_3, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 

0.8366532752144114

In [ ]:
rf_model_4 = RandomForestClassifier(n_estimators= 200, min_samples_split = 20 ,random_state = 42)

cross_val_score(rf_model_4, x_train, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 

0.8384520981358833

**위 실험을 통해 n_estimators(결정트리의 갯수)를 높게 지정할 수록 성능이 높아진다는 것을 알 수 있습니다. 또한 min_samples_split(노드를 분할하기 위한 최소한의 샘플 데이터수)를 높게 설정할 수록 성능이 좋아지는 결과를 확인할 수 있었습니다. n_estimators =200, min_samples_split = 20으로 설정한 rf_model_4 모델의 성능이 가장 뛰어났습니다. 하지만 n_estimators를 증가시킬 수록 실행시간이 오래걸리기 때문에 무작정 크게 증가시키는 것은 옳지 않습니다.**

### **Random Forest에 Feature selection 기법 적용**

In [ ]:
x_new = k_new(x_train, y_train)

cross_val_score(rf_model, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 

0.7517554108239675

In [ ]:
cross_val_score(rf_model_1, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 

0.752852286933512

In [ ]:
cross_val_score(rf_model_2, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 

0.7502196110264514

In [ ]:
cross_val_score(rf_model_3, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 

0.7908916789966777

In [ ]:
cross_val_score(rf_model_4, x_new, y_train, cv=5).mean()

/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/usr/local/lib/python3.8/dist-packages/sklearn/model_selection/_validation.py:680: DataConversionWarning: A column-vector y was passed when a 

0.8080031484237524

**Random Forest 모델에 Feature selection을 적용한 결과 모든 모델에서 score 값이 감소되었습니다. min_samples_split을 설정한 모델은 감소된 폭이 작았으며, default 값으로 설정한 모델은 큰 폭으로 성능이 감소했습니다. 이러한 이유를 추측해보자면, Decision Tree와 비슷한 결과로 생각해볼 수 있을 것 같습니다. Random Forest는 여러 개의 Decision Tree를 활용하여 배깅 방식을 사용합니다. 배깅방식을 이용하여 샘플데이터를 학습하고 투표하기 때문에 충분히 좋은 모델이 생성됩니다. 하지만 Feature selection을 진행하게 된다면 잠재적인 Feature에 대해서 놓칠 수 있기 때문에 오히려 Feature selection을 사용하지 않은 것이 모델의 성능에 득이 될것이라고 추측할 수 있습니다.**

### **Random Forest: test set에 대한 최종 평가**

In [ ]:
rf_model.fit(x_train, y_train)
rf_model.score(x_test, y_test)

<ipython-input-89-498991770b03>:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf_model.fit(x_train, y_train)


0.8175862421946976

In [ ]:
rf_model_1.fit(x_train, y_train)
rf_model_1.score(x_test, y_test)

<ipython-input-90-0dd7a9f16609>:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf_model_1.fit(x_train, y_train)


0.8182004299314157

In [ ]:
rf_model_2.fit(x_train, y_train)
rf_model_2.score(x_test, y_test)

<ipython-input-91-4c10dce822be>:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf_model_2.fit(x_train, y_train)


0.8077592384072064

In [ ]:
rf_model_3.fit(x_train, y_train)
rf_model_3.score(x_test, y_test)

<ipython-input-92-800a7a132298>:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf_model_3.fit(x_train, y_train)


0.8365236974101751

In [ ]:
rf_model_4.fit(x_train, y_train)
rf_model_4.score(x_test, y_test)

<ipython-input-93-bc6d6f2b2ab2>:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  rf_model_4.fit(x_train, y_train)


0.8397993653393387

**test set을 적용한 결과 train set을 적용한 값과 동일하게 rf_model_4 모델이 가장 좋은 성능을 보였습니다.n_estimators= 200, min_samples_split = 20으로 결정트리의 갯수를 늘리는 것은 성능이 좋아지는 것과 비례한다는 것을 확인할 수 있었습니다. 하지만 너무 큰 값을 설정하는 것은 실행 속도를 매우 느리게 만들기 때문에 적절한 결정트리 개수를 설정하는 것이 중요합니다. 또한 min_samples_split을 적절한 크기를 선택해 과적합 가능성을 줄이는 것이 모델의 성능에 도움이 된다는 것 또한 확인할 수 있었습니다. Random Forest 알고리즘은 결정트리를 통해 투표를 진행하기 때문에 결정트리의 값이 많은 것이 성능 향상에 영향을 미칩니다.**

# **결론**

**이번 실험을 통해 총 5가지의 classification 모델을 설계 해보았습니다.**

**Decision Tree Accuracy**: 0.8390228125014136

Decision Tree 모델은 데이터 수가 많을 때 좋은 모델입니다. 제가 선택한 Adult Data set은 데이터의 수가 상당히 많기 때문에 Decision Tree 모델이 두번째로 높은 정확도를 보여줄 수 있다고 생각합니다. 다른 모델들에 비해 실행시간이 빠르다는 점과 모델을 해석하는데 쉽다는 점이 매력적이였습니다. 

**Multi-Layer Perceptron Accuracy**: 0.8110349063363701

MLP 모델을 진행할때 수행시간이 매우 오래걸렸기 때문에 다른 모델에 비해 파라미터 변경을 많이 해보지 못한점이 아쉬운 점으로 남았습니다. 수행시간이 오래 걸린것에 비해 정확도가 낮기 때문에 MLP 모델은 사용하지 않을 것 같습니다. 하지만 많은 실험을 해보지 못했기 때문에 기회가 된다면 다른 실험에서 더욱 더 많은 연구를 해보고 싶습니다. 효율이 낮은 것에 대해 추측해보자면 MLP를 사용하기에 적당한 data set이 아니라고 생각되었습니다. MLP는 연속적 값들에 대한 예측을 할 때 유리하기 때문에 Decision Tree에 비해 효율이 낮다고 생각합니다.

**LogisiticRegression Accuracy**: 0.8022315487767427

이번 실험에서 가장 낮은 정확도를 보인 모델입니다. LogisiticRegression은 특정 클래스를 예측하는 것이 아닌 특정 클래스에 속할 확률을 계산합니다. LogisiticRegression은 다른 알고리즘에 비해 간단하기 때문에 이정도 정확도라면 효율적인 측면에서 좋다고 생각합니다. sklearn에서 threshold 값을 default 0.5로 설정하는 것을 변경하여 실험을 진행한다면 더 좋은 성능을 보여줄 수 있을거라 예측합니다.

**K-Nearest Neighbor Accuracy**: 0.8042788412324701

KNN 모델 또한 좋지 못한 정확도를 보여줬습니다. 간단하고 빠르다는 점이 좋지만 낮은 정확도를 보였으며 데이터가 많다면 분류 단계가 느리다는 점 때문에 사용을 해야하는지 고민이 될 것 같습니다. knn 모델 또한 data set이 알맞지 않았기 때문에 정확도가 낮을 것이라고 예상했습니다. knn에 적합한 data set에 실험을 진행하여 결과를 비교해 보고 싶습니다.

**Random Forest Accuracy**: 0.8397993653393387

Random Forest는 앙상블 기법 중 하나로 이번 실험에서 가장 높은 정확도를 보였습니다. 하지만 제가 예상했던 정확도보다 낮게 나온 것에 의문이 들었고, 그 이유에 대해서 추측해본 결과 알맞는 parameter를 찾지 못한 점과 최적의 Feature를 찾지 못한 것을 원인으로 생각했습니다. Feature selection 기법을 사용했을 때 정확도가 떨어진 것 처럼 알맞는 Feature selection 기법을 사용하여 최적의 Feature select 하게 된다면 정확도가 증가할 것이라고 생각합니다.
